Source/Research.cpp

```diff
@@ -0,0 +1,197 @@
+// Research.cpp
+#include "HFL.h"
+
+#pragma pack(push,1)
+struct OP2TechInfo
+{
+	int techID;
+	int category;
+	int techLevel;
+	int plymouthCost;
+	int edenCost;
+	int maxScientists;
+	int lab; // (basic = 1, standard = 2, advanced = 3)
+	int playerHasTech; // (bit vector, bit set if player has tech)
+	int numUpgrades; // (number of "UNIT_PROP"s specified in the tech file)
+	int numRequiredTechs; // (number of "REQUIRES"s specified in the tech file)
+	char* techName;
+	char* description;
+	char* teaser;
+	char* improveDesc;
+	int* requiredTechNum; // (pointer to array of techNums that are required for this tech)
+	void* upgrade; // (pointer to struct containing upgrade info)
+	int numDependentTech; // (number of techs that depend on this one) (below is a pointer to an int? list of this many elements)
+	int* dependentTechNum; // (pointer to array of techNums of techs that depend on this one)
+};
+
+
+// [Global] 0x56C230
+struct OP2Research
+{
+	int numTechs;
+	OP2TechInfo** techInfo;
+	int maxTechID; // [maxTechLevel * 1000 + 999]
+};
+
+#pragma pack(pop)
+
+OP2Research *researchObj;
+
+int Research::GetTechCount()
+{
+	if (!isInited)
+		return HFLNOTINITED;
+
+	return researchObj->numTechs;
+}
+TechInfo Research::GetTechInfo(int index)
+{
+	TechInfo info;
+	info.internalPtr = (void*)HFLNOTINITED;
+
+	if (!isInited)
+		return info;
+
+	if (index < 0 || index >= researchObj->numTechs)
+		return info;
+
+	info.internalPtr = researchObj->techInfo[index];
+	return info;
+}
+int Research::GetMaxTechID()
+{
+	if (!isInited)
+		return HFLNOTINITED;
+
+	return researchObj->maxTechID;
+}
+
+int TechInfo::IsValid()
+{
+	if (!isInited)
+		return HFLNOTINITED;
+
+	return (internalPtr != NULL);
```

Comment:
Newer code is starting to prefer `nullptr`. Older code that is still using `NULL` may trigger compiler warnings in some contexts (if such warnings are enabled).

Reply:
I agree we want to switch to nullptr. Since nullptr isn't called anywhere else in HFL yet, it doesn't make sense to require it before merging a pull request. Probably better to open an issue and work on converting the code to use nullptr.

Reply:
I decided to change this to nullptr. I'm a bit outdated. I haven't used C++ at work since 2009, so I didn't know about it until I saw it used in OP2Utility.

Reply:
Brett makes a good point. HFL doesn't yet make use of `nullptr`, though does use `NULL` in a number of places. We'll have to do a post merge pass to update the library.

Looks like this is where we're starting with the modernization.